### PR TITLE
Cwivagg/matomo 20220618

### PR DESCRIPTION
**Pull request by cwivagg**
### Matomo Role

Instructions: https://github.com/cwivagg/iiab/tree/cwivagg/matomo_20220618/roles/matomo#readme [revised from June]

Updates Adam's pull request https://github.com/iiab/iiab/pull/3253.

Context links: 

1. https://github.com/iiab/iiab/issues/1268

2. https://github.com/iiab/iiab/issues/1762

### Testing instructions

```
curl iiab.io/fast.txt | sudo bash -s 3304
cd /opt/iiab/iiab
sudo ./runrole matomo
```

### CODE REVIEW

File: roles/0-init/tasks/validate_vars.yml [hunk 1]
```diff
@@ -63,7 +63,7 @@
 #
 # 2020-11-04: Fix validation of 5 [now 4] core dependencies, for ./runrole etc
 
-- name: Set vars_checklist for 44 + 44 + 40 vars ("XYZ_install" + "XYZ_enabled" + "XYZ_installed") to be checked
+- name: Set vars_checklist for 45 + 45 + 41 vars ("XYZ_install" + "XYZ_enabled" + "XYZ_installed") to be checked
```
Comment (holta) [inline roles/0-init/tasks/validate_vars.yml]: @cwivagg can you update this 1 line?

```suggestion
- name: Set vars_checklist for 46 + 46 + 42 vars ("XYZ_install" + "XYZ_enabled" + "XYZ_installed") to be checked
```

Thanks if possible as another PR (#3302) just landed right in front of us :smile:

- https://github.com/iiab/iiab/pull/3302/files#diff-37a60cf03710ef308e5b429635f5eb815ab05b23850b94330c0c873036237381

File: roles/matomo/README.adoc [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+Below the "Visitors" button is a second button, "Behavior". Click on the "Pages" button after clicking "Behavior" and you can see the various pages that have been visited by your users. You may not see activity from the most recent day, since Matomo only updates its records once per day.
+
+=== IIAB Tips, Tricks, and Gotchas
+
+1. If your Internet In A Box setup is without power and Internet access, it will not be a ble to keep time correctly. This is okay! But it means that the time-of-visit information in Matomo will not be correct.
```
Comment (holta) [inline roles/matomo/README.adoc]:
```suggestion
1. If your Internet-in-a-Box setup is without power and Internet access, it will not be able to keep time correctly. This is okay! But it means that the time-of-visit information in Matomo will not be correct.
```

**Comment by holta** [comment]
Git is being difficult!

@cwivagg can you resolve the 1-liner by clicking "Resolve Conflicts" at the bottom of this (#3304) ticket?
(Or I can do it if you prefer!)

**Comment by holta** [comment]
@cwivagg Git is definitely torturing us.  Apologies.

One more try to set https://github.com/cwivagg/iiab/blob/cwivagg/matomo_20220618/roles/0-init/tasks/validate_vars.yml#L67 to "46 + 46 + 42" instead of "45 + 45 + 41" ?  (Or I can do it if you prefer!)

**Comment by holta** [comment]
Huge thanks @cwivagg

I'll try to recruit several testing volunteers so we can hopefully refine (as necessary?) and merge in the coming week or so!

> curl iiab.io/fast.txt | sudo bash -s 3304

CLARIF that if anybody's not intimately familiar with IIAB, this approach (below) is better!

```
curl iiab.io/install.txt | sudo bash -s 3304
```

**Comment by pickypet** [comment]
I tried to install on a fresh ubuntu 22.04 vm pc.

This was the response:

```
TASK [matomo : Configure Matomo to track IIAB] ************************************
fatal: [127.0.0.1]: FAILED! => {"cache_control": "no-store, must-revalidate", "changed": false, "connection": "close", "content_security_policy": "default-src 'self' 'unsafe-inline' 'unsafe-eval'; img-src 'self' 'unsafe-inline' 'unsafe-eval' data:;", "content_type": "text/html; charset=utf-8", "cookies": {}, "cookies_string": "", "date": "Tue, 12 Jul 2022 20:08:56 GMT", "elapsed": 0, "msg": "Status code was 200 and not [302]: OK (unknown bytes)", "redirected": false, "referrer_policy": "same-origin", "server": "nginx/1.18.0 (Ubuntu)", "status": 200, "transfer_encoding": "chunked", "url": "http://box.lan/matomo/index.php?action=firstWebsiteSetup&module=Installation", "x_matomo_request_id": "6af5a"}

PLAY RECAP ************************************************************************
127.0.0.1                  : ok=249  changed=129  unreachable=0    failed=1    skipped=70   rescued=0    ignored=3  
```

**Comment by holta** [comment]
> I tried to install on a fresh ubuntu 22.04 vm pc.

Thanks @pickypet !

FYI @cwivagg if you can investigate:

1) He ran `curl iiab.io/fast.txt | sudo bash -s 3304` then selected 0 (zero) for testing.
2) He edited `/etc/iiab/local_vars.yml` to set these 2 lines:

   ```
   matomo_install: True
   matomo_enabled: True
   ```

3) He ran `sudo iiab` to try to complete the install.

4) His [iiab-diagnostics](https://github.com/iiab/iiab/blob/master/scripts/iiab-diagnostics.README.md) are here: http://sprunge.us/ezhN5j?en

**Comment by holta** [comment]
@cwivagg if it helps to debug, the above Ubuntu Server 22.04 PC VM was created using command:

```
multipass launch 22.04 -c 2 -m 2G -d 25G
```

(In other words, 2 CPU's, 2GB RAM, 25GB disk.)

Let us know if other OS's need to be tested?  I ask especially because we have a lot of IIAB communities focusing on Linux Mint 21[*] whose Betas are in the "testing" folder here:

https://www.linuxmint.com/mirrors.php ([QA site](https://community.linuxmint.com/iso))

[*] It's based on Ubuntu 22.04 but with...

1) The "Cinnamon" graphical desktop on serious PC's, or...
2) The "MATE" desktop on lightweight PC's, or...
3) The "Xfce" rudimentary desktop which is NOT popular as it's too primitive for end users!

PS Legacy OS's like Debian 11 "Bullseye" are also hugely important of course, if you need that tested now or soon?  Reason: that's essentially what underlies Raspberry Pi OS on almost every Raspberry Pi (-:

- #3307

**Comment by cwivagg** [comment]
> I tried to install on a fresh ubuntu 22.04 vm pc.
> 
> This was the response:
> 
> ```
> TASK [matomo : Configure Matomo to track IIAB] ************************************
> fatal: [127.0.0.1]: FAILED! => {"cache_control": "no-store, must-revalidate", "changed": false, "connection": "close", "content_security_policy": "default-src 'self' 'unsafe-inline' 'unsafe-eval'; img-src 'self' 'unsafe-inline' 'unsafe-eval' data:;", "content_type": "text/html; charset=utf-8", "cookies": {}, "cookies_string": "", "date": "Tue, 12 Jul 2022 20:08:56 GMT", "elapsed": 0, "msg": "Status code was 200 and not [302]: OK (unknown bytes)", "redirected": false, "referrer_policy": "same-origin", "server": "nginx/1.18.0 (Ubuntu)", "status": 200, "transfer_encoding": "chunked", "url": "http://box.lan/matomo/index.php?action=firstWebsiteSetup&module=Installation", "x_matomo_request_id": "6af5a"}
> 
> PLAY RECAP ************************************************************************
> 127.0.0.1                  : ok=249  changed=129  unreachable=0    failed=1    skipped=70   rescued=0    ignored=3  
> ```

Thank you for testing it and finding this problem! My guess is that we may have messed up something with the way the time part of the form is being filled out, resulting in an inability to complete submission. I'm looking into it.

**Comment by holta** [comment]
@cwivagg can you click "Resolve Conflict" and change this to "44 + 44 + 40" ?

(Or I can if you prefer!  As we race towards the new school year, beginning by/around September in most countries!)

Related:

- PR #3314

**Comment by cwivagg** [comment]
Adam, I've done so, but I'm still unable to run my own branch due to a merge conflict. I'm assuming the problem is related? Is there a better way I can keep this line up to date?

**Comment by holta** [comment]
@cwivagg the comment should be `44 + 44 + 40` but again it's just a comment, no worries!

Currently it contains a small typo `44 + 44 + 0` here:
https://github.com/cwivagg/iiab/blob/cwivagg/matomo_20220618/roles/0-init/tasks/validate_vars.yml#L67

_Let's try to get this PR fully merged before August, if possible, so this frivolous side issue can be put behind us!_

**Comment by avni** [comment]
PR which has this comment fix here: https://github.com/cwivagg/iiab/pull/1

**Comment by holta** [comment]
> PR which has this comment fix here: [cwivagg#1](https://github.com/cwivagg/iiab/pull/1)

40 days + 40 nights until [IIAB's release](https://github.com/iiab/iiab/wiki/IIAB-8.0-Release-Notes) around end of August, for the new school year?

(If only Matomo could help IIAB schools graph + learn about the weather !  Possibly [Node-RED](https://github.com/iiab/iiab/tree/master/roles/nodered) is a better IIAB app for that?  During this complicated year especially...)

_"The food security crisis could kill more people than Covid has [in Africa especially] says Senegal minister at G-20"_
https://www.cnbc.com/2022/07/19/food-security-crisis-could-kill-more-people-than-covid-amadou-hott.html

**Comment by cwivagg** [comment]
> I tried to install on a fresh ubuntu 22.04 vm pc.
> 
> This was the response:
> 
> ```
> TASK [matomo : Configure Matomo to track IIAB] ************************************
> fatal: [127.0.0.1]: FAILED! => {"cache_control": "no-store, must-revalidate", "changed": false, "connection": "close", "content_security_policy": "default-src 'self' 'unsafe-inline' 'unsafe-eval'; img-src 'self' 'unsafe-inline' 'unsafe-eval' data:;", "content_type": "text/html; charset=utf-8", "cookies": {}, "cookies_string": "", "date": "Tue, 12 Jul 2022 20:08:56 GMT", "elapsed": 0, "msg": "Status code was 200 and not [302]: OK (unknown bytes)", "redirected": false, "referrer_policy": "same-origin", "server": "nginx/1.18.0 (Ubuntu)", "status": 200, "transfer_encoding": "chunked", "url": "http://box.lan/matomo/index.php?action=firstWebsiteSetup&module=Installation", "x_matomo_request_id": "6af5a"}
> 
> PLAY RECAP ************************************************************************
> 127.0.0.1                  : ok=249  changed=129  unreachable=0    failed=1    skipped=70   rescued=0    ignored=3  
> ```

Thank you again for testing this feature! As I guessed, it was related to the way that Matomo processes timezones. Apparently the timezones set by Multipass (which defaults to your area, mine was set to EST) do not match Matomo's expectations. I did not encounter this problem during development on VirtualBox because Virtualbox defaults to UTC. I've simply deleted the line of code fetching Ansible's datetime (which is the system time, without intervention), which allows Matomo to fetch the system time itself. I tested this on Multipass and VirtualBox with my own defaults, and, for fun, with Singapore. All of them worked.

In short, I believe the issue is now fixed. Thank you again for bringing it to myattention.

**Comment by holta** [comment]
@cwivagg this doesn't seem to install on our mainline OS (Raspberry Pi OS, e.g. on Raspberry Pi 4 and similar computers).

Can you tell why this is happening below?

> TASK [matomo : Configure Matomo to track IIAB] *********************************
fatal: [127.0.0.1]: FAILED! => {"cache_control": "no-store, must-revalidate", "changed": false, "connection": "close", "content_security_policy": "default-src 'self' 'unsafe-inline' 'unsafe-eval'; img-src 'self' 'unsafe-inline' 'unsafe-eval' data:;", "content_type": "text/html; charset=utf-8", "cookies": {}, "cookies_string": "", "date": "Thu, 28 Jul 2022 21:03:51 GMT", "elapsed": 2, "msg": "Status code was 200 and not [302]: OK (unknown bytes)", "redirected": false, "referrer_policy": "same-origin", "server": "nginx/1.18.0", "status": 200, "transfer_encoding": "chunked", "url": "http://box.lan/matomo/index.php?action=firstWebsiteSetup&module=Installation", "x_matomo_request_id": "0d4e9"}

Anything else I should try?

Context: The above error is the result of running `cd /opt/iiab/iiab` and then `sudo ./runrole matomo` on an existing IIAB (that was freshly installed 6 days ago).

**Comment by cwivagg** [comment]
> @cwivagg this doesn't seem to install on our mainline OS (Raspberry Pi OS, e.g. on Raspberry Pi 4 and similar computers).
> 
> Can you tell why this is happening below?
> 
> > TASK [matomo : Configure Matomo to track IIAB] *********************************
> > fatal: [127.0.0.1]: FAILED! => {"cache_control": "no-store, must-revalidate", "changed": false, "connection": "close", "content_security_policy": "default-src 'self' 'unsafe-inline' 'unsafe-eval'; img-src 'self' 'unsafe-inline' 'unsafe-eval' data:;", "content_type": "text/html; charset=utf-8", "cookies": {}, "cookies_string": "", "date": "Thu, 28 Jul 2022 21:03:51 GMT", "elapsed": 2, "msg": "Status code was 200 and not [302]: OK (unknown bytes)", "redirected": false, "referrer_policy": "same-origin", "server": "nginx/1.18.0", "status": 200, "transfer_encoding": "chunked", "url": "http://box.lan/matomo/index.php?action=firstWebsiteSetup&module=Installation", "x_matomo_request_id": "0d4e9"}
> 
> Anything else I should try?
> 
> Context: The above error is the result of running `cd /opt/iiab/iiab` and then `sudo ./runrole matomo` on an existing IIAB (that was freshly installed 6 days ago).

This looks like the same error. If you're sure you picked up the most recent commit, we can try two things:
1) I can get my own Raspberry Pi and try to reproduce;
2) I can give you a debugging branch that will write the output we need to a file, which you can then post back here.

1) might take a bit longer, but I can start looking into it today. For 2) [this branch](https://github.com/cwivagg/iiab/tree/cwivagg/matomo_20220618_debug) should capture the debug output in the directory from which execution is happening. In principle, the same result could be achieved by using Ansible's debug to print the output to the screen, but then there is a pretty hefty copy/paste job.

**Comment by holta** [comment]
> This looks like the same error. If you're sure you picked up the most recent commit,

Yes, fyi:

```
root@box:/opt/iiab/iiab# git log
commit 4e9635f97e97af549cea4cb0ee9d7b434b20cdd5 (HEAD -> matomo)
Merge: 2efd7c01 cd635851
Author: root <holta@users.noreply.github.com>
Date:   Thu Jul 28 21:01:21 2022 +0000

    Merge branch 'cwivagg/matomo_20220618' of https://github.com/cwivagg/iiab into matomo

commit cd6358514d0cab3e660aa916a7b6f5420888c0ac
Author: A Holt <holta@users.noreply.github.com>
Date:   Thu Jul 28 16:55:16 2022 -0400

    Modernize matomo/tasks/main.yml for skip_role_on_error #3255

commit 2efd7c01982ccecfb1297ae2bc93087172b38c2a (origin/master, origin/HEAD, master)
Merge: 004b8901 9ad5a9ca
Author: A Holt <holta@users.noreply.github.com>
Date:   Thu Jul 28 10:23:35 2022 -0400

    Merge pull request #3323 from holta/can_be_ap-patch

    network/tasks/detected_network.yml: can_be_ap grep failed on absence of regex

commit 9ad5a9caa128cbb5e3748fe6fe8f58f9619ff5b4
Author: A Holt <holta@users.noreply.github.com>
Date:   Thu Jul 28 14:18:49 2022 +0000

    network/tasks/detected_network.yml: can_be_ap grep failed on absence of regex

commit c6ee32d0c08bf7d6595d7c9b707d3f47352bc043
Author: Carl Wivagg <cwivagg@gmail.com>
Date:   Thu Jul 28 06:36:33 2022 -0400

    fix timezone issue
```

> we can try two things:
> 
> 1. I can get my own Raspberry Pi and try to reproduce;
> 2. I can give you a debugging branch that will write the output we need to a file, which you can then post back here.
> 3. might take a bit longer, but I can start looking into it today. For 2) [this branch](https://github.com/cwivagg/iiab/tree/cwivagg/matomo_20220618_debug) should capture the debug output in the directory from which execution is happening. In principle, the same result could be achieved by using Ansible's debug to print the output to the screen, but then there is a pretty hefty copy/paste job.

I'll give you direct/online access to several Raspberry Pi's.  Call me if poss?

PS I'm pretty sure this IIAB restructuring commit ("Modernize matomo/tasks/main.yml for skip_role_on_error #3255") is 100% irrelevant &mdash; but possibly pull/merge/cherry-pick this commit into your debug branch: https://github.com/iiab/iiab/pull/3304/commits/cd6358514d0cab3e660aa916a7b6f5420888c0ac

**Comment by holta** [comment]
@cwivagg call me if poss?  So you can log in and grab the debug info etc.

Yesterday's test had 64-bit Raspberry Pi OS Lite 2022-04-04 set to `Etc/UTC` (using this PR's branch [cwivagg/matomo_20220618](https://github.com/cwivagg/iiab/tree/cwivagg/matomo_20220618)) and today's new test used `America/New_York` with a freshly installed copy of the same OS (below, using test branch [cwivagg/matomo_20220618_debug](https://github.com/cwivagg/iiab/tree/cwivagg/matomo_20220618_debug)).

On this virgin OS today, I then ran as root:

```
mkdir /opt/iiab
cd /opt/iiab
apt install git
git clone https://github.com/iiab/iiab
cd /opt/iiab/iiab
git switch -c matomo    # Same as: git checkout -b matomo
git pull https://github.com/cwivagg/iiab.git cwivagg/matomo_20220618_debug
curl iiab.io/fast.txt | bash
```

Then I typed '0' [zero] for a [/etc/iiab/local_vars.yml tailored to unit-testing](https://github.com/iiab/iiab/blob/master/vars/local_vars_unittest.yml) which I then edited to contain:

```
matomo_install: True
matomo_enabled: True
```

Then I ran `sudo iiab` to try to complete the install.

Today's results are:

> TASK [matomo : Configure Matomo to track IIAB] *************************************************************************************************************************
fatal: [127.0.0.1]: FAILED! => {"changed": false, "connection": "close", "content": "", "content_type": "text/html; charset=UTF-8", "date": "Fri, 29 Jul 2022 17:41:18 GMT", "elapsed": 1, "location": "http://box.lan/matomo/index.php?action=trackingCode&module=Installation&site_idSite=1&site_name=IIAB", "msg": "Status code was 302 and not [200]: HTTP Error 302: Found", "redirected": false, "server": "nginx/1.18.0", "status": 302, "transfer_encoding": "chunked", "url": "http://box.lan/matomo/index.php?action=firstWebsiteSetup&module=Installation", "x_matomo_request_id": "2e174", "x_robots_tag": "noindex"}

**Comment by holta** [comment]
FYI Matomo 4.11 is now official!

https://github.com/matomo-org/matomo/releases/tag/4.11.0

**Comment by holta** [comment]
Thank you to @cwivagg who is diligently testing this PR on all 4 of IIAB's mainline OS's!

[1] http://FAQ.IIAB.IO -> "What OS should I use?"
[2] https://github.com/iiab/iiab/wiki/IIAB-Platforms

Let's get this merged soon after 6 months of hard work!

**Comment by holta** [comment]
_Huge Thanks @cwivagg for all your work, and for having tested this PR on 3+ different OS's (Mint 21, Ubuntu 22.04, Raspberry Pi OS Lite on Raspberry Pi 4) !_

Let's declare victory &mdash; adding any touch-ups later, in subsequent PR's &mdash; as grassroots communities begin to learn how to use this and then ask for refinements/assistance en route :100:
